### PR TITLE
feat: add test builds of the snap on pull_request events

### PR DESCRIPTION
**Pull request by ca-scribner**
Adds test builds to assert the snap is buildable before changes are merged into main.  This will help identify future build failures like when [this change was merged](https://github.com/canonical/grafana-agent-snap/actions/runs/8753095447) before merging to main, and will give [automated builds](https://github.com/canonical/grafana-agent-snap/pull/61) some CI to run against

OPENG-2388

### CODE REVIEW

**Review by sed-i — COMMENTED** [review]
Would be great to have this.
Could this be merged with the "release" CI?
Perhaps using `if event` conditionals, or calling a new, shared, `_build_snap` workflow?

**Comment by ca-scribner** [comment]
@sed-i I agree with you that it would be nice to deduplicate, but I'm not sure if the available options are better.

If we merge this into the "release" CI by adding `if` statements, would it feel strange that a `pull_request` event has a `release` CI run that passes but does not actually release anything? 

If we put the common parts in a reusable  `_build_snap` workflow, `_build_snap` would run as a separate job (eg: separate runners).  The publish CI needs the .snap artifact produced by `_build_snap`, so we'd need to include [upload-artifact](https://github.com/actions/upload-artifact?tab=readme-ov-file#actionsupload-artifact)/download-artifact steps to pass the artifact.  Since `_build_snap` only reuses ~15 lines of code, the new upload/download code will be about as much as we save.  So I'm not sure its worth it

tl;dr: I'm happy to do anything (all in release.yaml, make a _build_snap.yaml, or leave as is), but I don't have a strong preference on which :)  wdyt?

**Comment by ca-scribner** [comment]
I poked @lucabello about this offline and I think he's roughly in the same "we could do whatever" space.  He raised how snap building will sometime be replaced by a central build CI too, so whatever we do here will likely get overwritten soonish

**Comment by ca-scribner** [comment]
@sed-i and I discussed this offline.  We decided on leaving the duplication as is, but to add a comment at the top of the workflows so people know if they're editing one they should edit both.

File: .github/workflows/test-build-snap.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+          channel: latest/stable
+
+      - name: Install dependencies
+        run: |
+          sudo snap install --classic --channel edge snapcraft
+          # Setup Launchpad credentials
+          mkdir -p ~/.local/share/snapcraft
+          echo -e "$LAUNCHPAD_TOKEN" >> ~/.local/share/snapcraft/launchpad-credentials
+          git config --global user.email "github-actions@github.com"
+          git config --global user.name "Github Actions"
```
Comment (sed-i) [inline .github/workflows/test-build-snap.yaml:36]: Is git config needed for this workflow?

**Review by sed-i — APPROVED** [review]

File: .github/workflows/test-build-snap.yaml (hunk 1 above)
Comment (ca-scribner) [reply .github/workflows/test-build-snap.yaml:36]: I'm not sure tbh.  That is a good question both for this new workflow and the existing release.yaml.  I've captured that in #69 so we can land this feature and address that separately
